quic/s2n-quic-core/src/xdp/decoder.rs

```diff
@@ -3,83 +3,143 @@
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_ipv6_header(&mut self, header: &ipv6::Header) -> Result {
+        let _ = header;
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_udp_header(&mut self, header: &udp::Header) -> Result {
+        let _ = header;
+        Ok(Some(()))
     }
 }
 
-impl Validator for () {}
+impl EventHandler for () {}
+
+impl EventHandler for path::Tuple {
+    #[inline(always)]
+    fn on_ethernet_header(&mut self, header: &ethernet::Header) -> Result {
+        self.remote_address.mac = header.source;
+        self.local_address.mac = header.destination;
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_ipv4_header(&mut self, header: &ipv4::Header) -> Result {
+        self.remote_address.ip = header.source.into();
+        self.local_address.ip = header.destination.into();
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_ipv6_header(&mut self, header: &ipv6::Header) -> Result {
+        self.remote_address.ip = header.source.into();
+        self.local_address.ip = header.destination.into();
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_udp_header(&mut self, header: &udp::Header) -> Result {
+        self.remote_address.port = header.source.get();
+        self.local_address.port = header.destination.get();
+        Ok(Some(()))
+    }
+}
+
+impl<P: EventHandler> EventHandler for datagram::Header<P> {
+    #[inline(always)]
+    fn on_ethernet_header(&mut self, header: &ethernet::Header) -> Result {
+        self.path.on_ethernet_header(header)
+    }
+
+    #[inline(always)]
+    fn on_ipv4_header(&mut self, header: &ipv4::Header) -> Result {
+        self.path.on_ipv4_header(header)?;
+        self.ecn = header.tos().ecn();
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_ipv6_header(&mut self, header: &ipv6::Header) -> Result {
+        self.path.on_ipv6_header(header)?;
+        self.ecn = header.vtcfl().ecn();
+        Ok(Some(()))
+    }
+
+    #[inline(always)]
+    fn on_udp_header(&mut self, header: &udp::Header) -> Result {
+        self.path.on_udp_header(header)
+    }
+}
 
 /// Decodes a path tuple and payload from a raw packet
 #[inline(always)]
-pub fn decode_packet<'a, D: Decoder<'a>>(buffer: D) -> Result<path::Tuple, D> {
-    decode_packet_validator(buffer, &())
+pub fn decode_packet<'a, D: Decoder<'a>>(
+    buffer: D,
+) -> core::result::Result<Option<(datagram::Header<path::Tuple>, D)>, DecoderError> {
+    let mut header = datagram::Header {
+        path: path::Tuple::UNSPECIFIED,
+        ecn: Default::default(),
+    };
+    match decode_packet_with_event(buffer, &mut header)? {
+        Some(buffer) => Ok(Some((header, buffer))),
+        None => Ok(None),
+    }
 }
 
 /// Decodes a path tuple and payload from a raw packet
 #[inline(always)]
-pub fn decode_packet_validator<'a, D: Decoder<'a>, V: Validator>(
+pub fn decode_packet_with_event<'a, D: Decoder<'a>, E: EventHandler>(
```

Comment:
ahh.. took me a minute to realize that this is replacing Validator and that events are being used for validation now. maybe renaming `with_event` to `with_validation`. Non-blocking.

Reply:
Yeah I wanted to make it more generic than _just_ validation, since we're also using it for header field extraction (IP, port, ECN, etc.)